comms/storage/longterm/longterm.go

```diff
@@ -0,0 +1,158 @@
+		return nil, fmt.Errorf("failed to read %q: %v", key, err)
+	}
+	return reader, nil
+}
+
+// runStorer runs a goroutine to pull tracks from temp NATS object storage to long-term object storage.
+func (lt *LongTerm) runStorer(uploadStream string) {
+	thisNodePubKey := os.Getenv("audius_delegate_owner_wallet") // TODO: Get from config or something - same for value in NewProd() above
+	// Create a per-node explicit pull consumer on the stream that backs the track upload status KV bucket
+	storerDurable := fmt.Sprintf("STORER_%s", thisNodePubKey)
+	_, err := lt.jsc.AddConsumer(uploadStream, &nats.ConsumerConfig{
+		Durable:       storerDurable,
+		AckPolicy:     nats.AckExplicitPolicy,
+		DeliverPolicy: nats.DeliverAllPolicy, // Using the "all" policy means when a node registers it will download every track that it needs
+		ReplayPolicy:  nats.ReplayInstantPolicy,
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	// Create a subscription on the consumer for every node
+	// Subject can be empty since it defaults to all subjects bound to the stream
+	storerSub, err := lt.jsc.PullSubscribe("", storerDurable, nats.BindStream(uploadStream))
+	if err != nil {
+		panic(err)
+	}
+
+	// Watch KV store to download files to long-term storage
+	// TODO: Maybe there should be an exit channel for in case we restart StorageServer without restarting the whole program? (e.g., if we want to update StorageDecider to pass it a new slice of storage node pubkeys)
+	go func() {
+		for {
+			msgs, err := storerSub.Fetch(1)
+			if err == nil {
+				msgs[0].Ack()
+
+				job := transcode.Job{}
+				err := json.Unmarshal(msgs[0].Data, &job)
+				if err != nil {
+					panic(err)
+				}
+
+				if job.Status == transcode.JobStatusDone {
+					if lt.storageDecider.ShouldStore(job.ID) {
+						fmt.Printf("Storing file with ID %q\n", job.ID)
+						lt.moveTempToLongTerm(job.Results)
+					} else {
+						fmt.Printf("Not storing file with ID %q\n", job.ID)
+						continue
+					}
+				}
+			} else if err != nats.ErrTimeout { // Timeout is expected when there's nothing new in the stream
+				fmt.Printf("Error fetching message to store a file: %q\n", err)
+			}
+		}
+	}()
+}
+
+func (lt *LongTerm) moveTempToLongTerm(tmpObjects []*nats.ObjectInfo) {
+	// Open the long-term storage *blob.Bucket
+	ctx := context.Background()
+	bucket, err := blob.OpenBucket(ctx, getBucketUrl())
+	if err != nil {
+		fmt.Printf("error: failed to store file - could not open bucket: %v", err)
+		return
+	}
+	defer bucket.Close()
+
+	for _, tmpObj := range tmpObjects {
+		// Get object from temp store
+		objStore, err := lt.jsc.ObjectStore(tmpObj.Bucket)
+		if err != nil {
+			log.Printf("Failed to get object from temp store: %v\n", err)
+			return
+		}
+		tempObj, err := objStore.Get(tmpObj.Name)
+		if err != nil {
+			log.Printf("Failed to get object from temp store: %v\n", err)
+			return
+		}
+
+		// Open a *blob.Writer for the blob at key=tmpObj.Name
+		writer, err := bucket.NewWriter(ctx, tmpObj.Name, nil)
```

Comment:
Two things to note here:

1. We had talked about "each shard has a s3 bucket"... but really there should just be one bucket and the key should be like: `/abc/Qm53abc` where `abc` is the suffix we use for sharding.  This way we just create one s3 bucket, but still use `s3 sync --include /abc/*`.  Not sure if CDK file driver will create intermediate folders... but I bet it will!

3. If the input hashes to `Qm53abc` the result will be something like `Qm53abc_150x.jpg` with a suffix variant... at some point we'll have to look at what the client uses and adjust the result naming to match.

Reply:
1. good point! since it seemed both future-proof and easier to implement, I changed it to use the temp obj store's bucket name, which is `<namespace>_<bucket>`. so in long-term storage it's like `0_abc/Qm53abc_150x.jpg` and we can still use `s3 sync --include 0_abc/*` like suggested. also, you're right that it is indeed creating folders - see screenshot:
<img width="691" alt="Screen Shot 2023-02-02 at 11 22 34 AM" src="https://user-images.githubusercontent.com/4657956/216429492-c71ac3ea-997c-48b9-9272-c89699aa23b8.png">

2. I'm making a Linear ticket with some open questions to also include some options you mentioned earlier today